### DISCUSSION TITLE
Pydantic V2: `Enum` as `dict` key: UserWarning (serialization)

### DISCUSSION THREAD

**Discussion by ames0k0** [fastapi/fastapi#10130] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum
from pydantic import BaseModel, Field, ConfigDict
from fastapi import FastAPI

data = { 'fruits': {'apple': 1} }

class Fruit(str, Enum):
    apple = 'apple'


class Bar(BaseModel):
    fruits: dict[Fruit, int] = Field(..., description='Fruits count')


Bar.model_validate(data)
print('[!] valid')


app = FastAPI()


@app.post('/', response_model=Bar)
async def foo() -> dict:
    return data
```

### Description

[!] valid
INFO:     Started server process [378388]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
/src/env/lib/python3.10/site-packages/pydantic/type_adapter.py:268: UserWarning: Pydantic serializer warnings:
  Expected `json-or-python[json=str, python=any]` but got `Fruit` - serialized value may not be as expected
  return self.serializer.to_python(
INFO:     127.0.0.1:33528 - "POST / HTTP/1.1" 200 OK

### Operating System

Linux

### Operating System Details

Ubuntu 20.04.6

### FastAPI Version

0.101.1

### Pydantic Version

2.2.1

### Python Version

Python 3.10.12

### Additional Context

Getting this warning after update, works fine for `fastapi==0.99`

```bash
uvicorn main:app --reload
```

**Comment by JavierSanchezCastro** [comment, marked as answer]
I think it's pydactic side, check that issue

https://github.com/pydantic/pydantic/issues/2278

**Reply by ames0k0** [reply]
Thank you very much!

**Comment by ames0k0** [comment]
Sorry for my bad research!
Here's updated code
```python
class Bar(BaseModel):
    fruits: dict[Fruit, int] = Field(..., description='Fruits count')

    model_config = ConfigDict(json_encoders={Enum: lambda v: v.value})
```
